src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -397,42 +403,65 @@ internal void Assign(ICollection<TopicPartitionOffset> partitions)
             }
             if (err != ErrorCode.NO_ERROR)
             {
-                throw new KafkaException(err, "Failed to assign partitions");
+                throw new KafkaException(err);
             }
         }
 
-        internal void Commit()
+        /// <summary>
+        ///  Dummy commit callback that does nothing but prohibits
+        ///  triggering the global offset_commit_cb.
+        ///  Used by manual commits.
+        /// </summary>
+        static void dummyOffsetCommitCb (IntPtr rk, ErrorCode err, IntPtr offsets, IntPtr opaque)
         {
-            ErrorCode err = LibRdKafka.commit(handle, IntPtr.Zero, false);
-            if (err != ErrorCode.NO_ERROR)
-            {
-                throw new KafkaException(err, "Failed to commit offsets");
-            }
+            return;
         }
 
-        internal void Commit(ICollection<TopicPartitionOffset> offsets)
+        /// <summary>
+        /// Manual sync commit, will block indefinately.
+        /// </summary>
+        /// <param name="offsets">Offsets to commit, or null for current assignment.</param>
+        /// <returns>CommittedOffsets with global or per-partition errors.</returns>
+        private CommittedOffsets commitSync (ICollection<TopicPartitionOffset> offsets)
         {
-            IntPtr list = LibRdKafka.topic_partition_list_new((IntPtr) offsets.Count);
-            if (list == IntPtr.Zero)
-            {
-                throw new Exception("Failed to create offset commit list");
-            }
-            foreach (var offset in offsets)
+            // Create temporary queue so we can get the offset commit results
+            // as an event instead of a callback.
+            // We still need to specify a dummy callback (that does nothing)
+            // to prevent the global offset_commit_cb to kick in.
+            IntPtr cQueue = LibRdKafka.queue_new(handle);
```

Comment:
I still don't properly understand whether `SafeHandle` has any meaningful advantages in situations like this (where the `IntPtr` is just some memory given to us by librdkafka). Whether or not it does, we can't use it to own the pointer because commit_queue takes ownership (conditional on success/failure).